### ISSUE TITLE
Generic ParamSpec in union causes weird invariant

### ISSUE THREAD

**Issue by Bluenix2** [microsoft/pyright#2364]
**Describe the bug**
I have a union of generics with a ParamSpec and TypeVar (the latter for the return type) but when passing an instance of one of these types PyRight produces a weird error about invariance.

**To Reproduce**
I have created this reproduceable code:
```python
from typing import Generic, TypeVar, Union
from typing_extensions import ParamSpec

P = ParamSpec('P')
RT = TypeVar('RT')


class Foo(Generic[P, RT]):
    ...


class Bar(Generic[P, RT]):
    ...


FooBar = Union[Foo[P, RT], Bar[P, RT]]


def something(item: FooBar) -> List[FooBar]:
    return [item]

something(Foo())
```
```
Argument of type "Foo[P@Foo, Unknown]" cannot be assigned to parameter "item" of type "FooBar[Unknown, Unknown]" in function "something"
  Type "Foo[P@Foo, Unknown]" cannot be assigned to type "FooBar[Unknown, Unknown]"
    TypeVar "P@Foo" is invariant
      Type "P@Foo" cannot be assigned to type "P@FooBar"
    "Foo[P@Foo, Unknown]" is incompatible with "Bar[P@FooBar, RT@FooBar]"
```
**Expected behavior**
To me this should pass, I don't see what would be wrong about this.

**VS Code extension or command-line**
I am using the most recent version as far as I am aware:
```
Released on
22/03/2019, 04:10:23
Last updated
02/10/2021, 01:34:43
Identifier
ms-pyright.pyright
```

**Comment by erictraut** [comment]
Thanks for filing the issue.

There are a couple of problems with your code sample.

First, the generic type alias `FooBar` requires two type arguments (the first of which must be a ParamSpec), but you are not supplying any type arguments when you use `FooBar` in the `something` function. It should look like this:
```python
def something(item: FooBar[P, RT]) -> List[FooBar[P, RT]]:
    return [item]
```

If you don't provide type arguments, they are assumed to be `Any`. Pyright uses a special form of `Any` called `Unknown` to signify an implicit `Any` type, so the type of `FooBar` is interpreted as `FooBar[Unknown, Unknown]`.

If you enable the `reportMissingTypeArgument` diagnostic check, pyright will flag missing type arguments for you.

Second, the `Foo` class is generic, but it has no constructor that provides type arguments for its type parameters when it is instantiated. Perhaps you omitted the constructor to simplify the the sample. Without a constructor, the type of the expression `Foo()` is `Foo[Unknown, Unknown]`. That's probably not what you intend.

Those problems notwithstanding, there is a bug here in pyright. It should allow a value of type `Foo[Unknown, Unknown]` to be assigned to a parameter of type `FooBar[Unknown, Unknown]`. A fix for this will be in the next release.

In the meantime, you can work around the issue by fixing the two problems that I noted above.

**Comment by Bluenix2** [comment]
> There are a couple of problems with your code sample.
> 
> First, the generic type alias `FooBar` requires two type arguments (the first of which must be a ParamSpec), but you are not supplying any type arguments when you use `FooBar` in the `something` function. It should look like this:
> 
> ```python
> def something(item: FooBar[P, RT]) -> List[FooBar[P, RT]]:
>     return [item]
> ```

I tried to use `List[Foobar]` to better reflect more of my usecase. To give some more information you need a third `Container` type like this:
```python
import random
from typing import Callable, Generic, List, Optional, TypeVar, Union, overload
from typing_extensions import ParamSpec

P = ParamSpec('P')
RT = TypeVar('RT')


class Foo(Generic[P, RT]):
    func: Callable[P, RT]

    def __init__(self, func: Callable[P, RT]):
        self.func = func

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> RT:
        return self.func(*args, **kwargs)


class Bar(Generic[P, RT]):
    func: Callable[P, RT]

    def __init__(self, func: Callable[P, RT]):
        self.func = func

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> RT:
        return self.func(*args, **kwargs)


FooBar = Union[Foo[P, RT], Bar[P, RT]]


class Container:
    # I can either choose between making this `List[FooBar[P, RT]]` and get
    # a complaint about unnecessary ParamSpec and TypeVar
    _values: List[FooBar]

    def __init__(self) -> None:
        self._values = []

    def register(self, item: FooBar[P, RT]) -> None:
        # Or I can get an error here which I do with this code (1)
        self._values.append(item)

    @overload
    def stateful(self) -> Callable[[Callable[P, RT]], FooBar[P, RT]]: ...

    @overload
    def stateful(self, func: Callable[P, RT]) -> FooBar[P, RT]: ...

    def stateful(  # NEW ERROR (2)
        self,
        func: Optional[Callable[P, RT]] = None,
    ) -> Union[FooBar[P, RT], Callable[[Callable[P, RT]], FooBar[P, RT]]]:
        def decorator(wrapped: Callable[P, RT]) -> FooBar[P, RT]:
            if random.randint(0, 1) < 0.5:
                value = Foo(wrapped)
            else:
                value = Bar(wrapped)

            self.register(value)
            return value  # NEW ERROR (3)

        if callable(func):
            return decorator(func)

        return decorator


instance = Container()


@instance.stateful()
def example(arg: int) -> int:
    return arg
```
The first error I get is this (1):
```
Argument of type "FooBar[P@register, RT@register]" cannot be assigned to parameter "__object" of type "Foo[P@FooBar, RT@FooBar] | Bar[P@FooBar, RT@FooBar]" in function "append"
  Type "FooBar[P@register, RT@register]" cannot be assigned to type "Foo[P@FooBar, RT@FooBar] | Bar[P@FooBar, RT@FooBar]"
    Type "Foo[P@register, RT@register]" cannot be assigned to type "Foo[P@FooBar, RT@FooBar] | Bar[P@FooBar, RT@FooBar]"
      TypeVar "P@Foo" is invariant
        Type "P@register" cannot be assigned to type "P@FooBar"
      "Foo[P@register, RT@register]" is incompatible with "Bar[P@FooBar, RT@FooBar]"
    Type "Bar[P@register, RT@register]" cannot be assigned to type "Foo[P@FooBar, RT@FooBar] | Bar[P@FooBar, RT@FooBar]"
      "Bar[P@register, RT@register]" is incompatible with "Foo[P@FooBar, RT@FooBar]"
      TypeVar "P@Bar" is invariant
```
But then I get these two errors in my simplified example I cannot manage to replicate in my real implementation.  I suspect that this is to do with my real implementation having another member in the union (let's call it `Baz`) that shares a parent with `Bar`.

This is new (2):
```
Overloaded function implementation is not consistent with signature of overload 2
  Function return type "FooBar[P@stateful, RT@stateful]" is incompatible with type "Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful] | (_p0: Callable[P@stateful, RT@stateful]) -> (Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful])"
    Type "FooBar[P@stateful, RT@stateful]" cannot be assigned to type "Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful] | (_p0: Callable[P@stateful, RT@stateful]) -> (Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful])"
      Type "Foo[P@stateful, RT@stateful]" cannot be assigned to type "Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful] | (_p0: Callable[P@stateful, RT@stateful]) -> (Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful])"
        TypeVar "P@Foo" is invariant
          "object*" is incompatible with "type*"
        "Foo[P@stateful, RT@stateful]" is incompatible with "Bar[P@stateful, RT@stateful]"
        Type "Foo[P@stateful, RT@stateful]" cannot be assigned to type "(_p0: Callable[P@stateful, RT@stateful]) -> (Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful])"
          Parameter 1: type "Callable[P@stateful, RT@stateful]" cannot be assigned to type "P.args"
```
As well as (3):
```
Expression of type "Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful]" cannot be assigned to return type "FooBar[P@stateful, RT@stateful]"
  Type "Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful]" cannot be assigned to type "FooBar[P@stateful, RT@stateful]"
    Type "Foo[P@stateful, RT@stateful]" cannot be assigned to type "FooBar[P@stateful, RT@stateful]"
      TypeVar "P@Foo" is invariant
        Type "P@stateful" cannot be assigned to type "P@stateful"
      "Foo[P@stateful, RT@stateful]" is incompatible with "Bar[P@stateful, RT@stateful]"
    Type "Bar[P@stateful, RT@stateful]" cannot be assigned to type "FooBar[P@stateful, RT@stateful]"
      "Bar[P@stateful, RT@stateful]" is incompatible with "Foo[P@stateful, RT@stateful]"
      TypeVar "P@Bar" is invariant
  ...
```
This appears to cause more errors than I have in my implementation (which you [can find here if you really want to have a look](https://github.com/Bluenix2/wumpy/blob/main/wumpy/interactions/commands/registrar.py) just remove the `type: ignore` comments).


> Second, the `Foo` class is generic, but it has no constructor that provides type arguments for its type parameters when it is instantiated. Perhaps you omitted the constructor to simplify the the sample. Without a constructor, the type of the expression `Foo()` is `Foo[Unknown, Unknown]`. That's probably not what you intend.

I added the decorator to show a "stateful" example where the generics receives a type.

> Those problems notwithstanding, there is a bug here in pyright. It should allow a value of type `Foo[Unknown, Unknown]` to be assigned to a parameter of type `FooBar[Unknown, Unknown]`. A fix for this will be in the next release.
> 
> In the meantime, you can work around the issue by fixing the two problems that I noted above.

I am not sure if - in it's current state - I can implement those but I see.

I installed PyRight from npm using: `npm install git+https://github.com/microsoft/pyright.git` and ran `npx pyright` but still got the following errors from the codeblock above that I keep in `test.py`:
```python
C:\Users\%username%\Projects\wumpy\test.py
  C:\Users\%username%\Projects\wumpy\test.py:39:29 - error: Argument of type "FooBar[P@register, RT@register]" cannot be assigned to parameter "__object" of type "Foo[P@FooBar, RT@FooBar] | Bar[P@FooBar, RT@FooBar]" in function "append"
    Type "FooBar[P@register, RT@register]" cannot be assigned to type "Foo[P@FooBar, RT@FooBar] | Bar[P@FooBar, RT@FooBar]"
      Type "Foo[P@register, RT@register]" cannot be assigned to type "Foo[P@FooBar, RT@FooBar] | Bar[P@FooBar, RT@FooBar]"
        TypeVar "P@Foo" is invariant
          Type "P@register" cannot be assigned to type "P@FooBar"
        "Foo[P@register, RT@register]" is incompatible with "Bar[P@FooBar, RT@FooBar]"
      Type "Bar[P@register, RT@register]" cannot be assigned to type "Foo[P@FooBar, RT@FooBar] | Bar[P@FooBar, RT@FooBar]"
        "Bar[P@register, RT@register]" is incompatible with "Foo[P@FooBar, RT@FooBar]"
        TypeVar "P@Bar" is invariant (reportGeneralTypeIssues)
  C:\Users\%username%\Projects\wumpy\test.py:58:20 - error: Expression of type "Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful]" cannot be assigned to return type "FooBar[P@stateful, RT@stateful]"
    Type "Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful]" cannot be assigned to type "FooBar[P@stateful, RT@stateful]"
      Type "Foo[P@stateful, RT@stateful]" cannot be assigned to type "FooBar[P@stateful, RT@stateful]"
        TypeVar "P@Foo" is invariant
          Type "P@stateful" cannot be assigned to type "P@stateful"
        "Foo[P@stateful, RT@stateful]" is incompatible with "Bar[P@stateful, RT@stateful]"
      Type "Bar[P@stateful, RT@stateful]" cannot be assigned to type "FooBar[P@stateful, RT@stateful]"
        "Bar[P@stateful, RT@stateful]" is incompatible with "Foo[P@stateful, RT@stateful]"
        TypeVar "P@Bar" is invariant
    ... (reportGeneralTypeIssues)
  C:\Users\%username%\Projects\wumpy\test.py:47:9 - error: Overloaded function implementation is not consistent with signature of overload 2
    Function return type "FooBar[P@stateful, RT@stateful]" is incompatible with type "Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful] | (_p0: Callable[P@stateful, RT@stateful]) -> (Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful])"
      Type "FooBar[P@stateful, RT@stateful]" cannot be assigned to type "Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful] | (_p0: Callable[P@stateful, RT@stateful]) -> (Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful])"
        Type "Foo[P@stateful, RT@stateful]" cannot be assigned to type "Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful] | (_p0: Callable[P@stateful, RT@stateful]) -> (Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful])"
          TypeVar "P@Foo" is invariant
            "object*" is incompatible with "type*"
          "Foo[P@stateful, RT@stateful]" is incompatible with "Bar[P@stateful, RT@stateful]"
          Type "Foo[P@stateful, RT@stateful]" cannot be assigned to type "(_p0: Callable[P@stateful, RT@stateful]) -> (Foo[P@stateful, RT@stateful] | Bar[P@stateful, RT@stateful])"
            Parameter 1: type "Callable[P@stateful, RT@stateful]" cannot be assigned to type "P.args" (reportGeneralTypeIssues)
```
*Note: `test.py` isn't pushed to my repository, you can just copy the contents from this comment*

**Comment by erictraut** [comment]
Your new sample also contains the same problem that I mentioned above. `FooBar` is a generic type alias, so anywhere you use it, you should be supplying two type arguments to it. But your sample uses it without type arguments in the `_values` type annotation. It also introduces `P` and `RT` in many different scopes, so each of them are distinct. For example, `P` and `RT` are scoped to the `register` method and independently scoped to the `stateful` method and again independently in the `FooBar` type alias. All of these uses of `P` and `RT` are completely distinct, as though they didn't share a name. That's because of the scoping rules defined in PEP 484 for TypeVars.

I think your intent is to make the `Container` class generic based on the types `P` and `RT`. That way, all references to these two TypeVars within that class will refer to the same types.

```python
class Container(Generic[P, RT]):
    _values: List[FooBar[P, RT]]
   ...
```

Incidentally, you cannot install the latest version of pyright by simply cloning its repo. It needs to be rebuilt, since it's written in a compiled language (TypeScript). I recommend waiting until the next time I publish it to the npm registry and installing it from there.

**Comment by erictraut** [comment]
Here's a modified approach that uses a TypeVar bound to a Callable type.

```python
TCallback = TypeVar("TCallback", bound=Callable[..., Any])

class Container(Generic[TCallback]):
    _values: List[TCallback]

    def __init__(self) -> None:
        self._values = []

    def register(self, item: TCallback) -> None:
        self._values.append(item)

    @overload
    def stateful(self) -> Callable[[TCallback], TCallback]:
        ...

    @overload
    def stateful(self, func: TCallback) -> TCallback:
        ...

    def stateful(
        self,
        func: Optional[TCallback] = None,
    ) -> Union[TCallback, Callable[[TCallback], TCallback]]:
        def decorator(wrapped: TCallback) -> TCallback:
            if random.randint(0, 1) < 0.5:
                value = Foo(wrapped)
            else:
                value = Bar(wrapped)

            self.register(value)
            return cast(TCallback, value)

        if callable(func):
            return decorator(func)

        return decorator


instance = Container[Callable[[int], int]]()

@instance.stateful()
def example(arg: int) -> int:
    return arg
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.174, which I just published. It will also be included in the next release of pylance.

**Comment by Bluenix2** [comment]
Thank you very much for your help.
